core/client/client.go

```diff
@@ -151,6 +151,8 @@ func min(x, y int32) int32 {
 // ListHistory returns a list of profiles starting and ending at given epochs.
 // It also filters out all identical consecutive profiles.
 func (c *Client) ListHistory(ctx context.Context, userID, appID string, start, end int64, opts ...grpc.CallOption) (map[*types.MapRootV1][]byte, error) {
+	c.trustedLock.Lock()
```

Comment:
Since you acquire the lock here and don't release it, `c.trusted` will not change, so you can move `trustedSnapshot := c.trusted` out from the loop below. Alternatively, if you want to allow `c.trusted` updates happening in between loop iterations, please move this locking to the beginning of the loop.

Reply:
Also, consider using `RWMutex` instead of `Mutex`.

Reply:
Excellent observation.  Done. 
This GetHistory function needs a refactor which I'll do in a separate PR.